{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"from Bio.SubsMat import MatrixInfo\nimport numpy as np\nimport matplotlib.pyplot as plt","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Task 1\n\n**Write a program to implement Needleman-Wunsch for proteins.**"},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"# Pretty print matrices\ndef print_matrix(mat):\n\n    for i in range(0, len(mat)):\n        print(\"[\", end = \"\")\n        \n        for j in range(0, len(mat[i])):\n\n            print(mat[i][j], end = \"\")\n\n            if j != len(mat[i]) - 1:\n                print(\"\\t\", end = \"\")\n                \n        print(\"]\\n\")","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Make an empty matrix of 0s\ndef zeros(rows, cols):\n    retval = []\n\n    for x in range(rows):\n        retval.append([])\n\n        for y in range(cols):\n            retval[-1].append(0)\n            \n    return retval","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"We're using the blosum50 scoring matrix.\n\n<img src=\"https://i.imgur.com/IsXqXrK.png\" width=\"500\">"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Blosum 50\ndef blosum(alpha, beta):\n    \n    pair = (alpha, beta)\n    b50 = MatrixInfo.blosum50\n    \n    if pair not in b50:\n        return b50[tuple(reversed(pair))]\n    else:\n        return b50[pair]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Determinines the score between any two bases in alignment\ndef match_score(alpha, beta):\n    \n    if alpha == beta:\n        return blosum(alpha, beta)\n    elif alpha == '-' or beta == '-':\n        return gap_penalty\n    else:\n        return blosum(alpha, beta)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The Needleman-Wunsch algorithm uses the following dynamic programming formula for global alignment:\n\n<img src=\"https://i.imgur.com/tQp3P3J.png\" width=\"300\">\n\nWhere we have two strings \n$a = a_1 a_2 ... a_m$,\n$b = b_1 b_2 ... b_n$\n\nThe price of a substitution \n$s(a, b)$\n\nAnd the price of an insertion or deletion \n$d = âˆ’8$"},{"metadata":{"trusted":true},"cell_type":"code","source":"gap_penalty = -8","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def needleman_wunsch(seq1, seq2):\n    \n    n = len(seq1)  \n    m = len(seq2)\n    \n    # Initiate empty scoring matrix\n    score = zeros(m+1, n+1)\n    \n    # Fill out first column\n    for i in range(0, m + 1):\n        score[i][0] = gap_penalty * i\n    \n    # Fill out first row\n    for j in range(0, n + 1):\n        score[0][j] = gap_penalty * j\n    \n    # Fill out all other values in the score matrix\n    for i in range(1, m + 1):\n        for j in range(1, n + 1):\n            \n            # Calculate the score by checking the top, left, and diagonal cells\n            match = score[i - 1][j - 1] + match_score(seq1[j-1], seq2[i-1])\n            delete = score[i - 1][j] + gap_penalty\n            insert = score[i][j - 1] + gap_penalty\n            \n            # Record the maximum score from the three possible scores calculated above\n            score[i][j] = max(match, delete, insert)\n    \n    #print_matrix(score)\n    \n    align1 = \"\"\n    align2 = \"\"\n    \n    # Start from the bottom right cell in matrix\n    i = m\n    j = n\n    \n    # We finish this loop when we reach the top or the left edge\n    while i > 0 and j > 0:\n        score_current = score[i][j]\n        score_diagonal = score[i-1][j-1]\n        score_up = score[i][j-1]\n        score_left = score[i-1][j]\n        \n        # Figure out which cell the score came from\n        if score_current == score_diagonal + match_score(seq1[j-1], seq2[i-1]):\n            align1 += seq1[j-1]\n            align2 += seq2[i-1]\n            i -= 1\n            j -= 1\n        elif score_current == score_up + gap_penalty:\n            align1 += seq1[j-1]\n            align2 += '-'\n            j -= 1\n        elif score_current == score_left + gap_penalty:\n            align1 += '-'\n            align2 += seq2[i-1]\n            i -= 1\n\n    # Finish tracing up to the top left corner\n    while j > 0:\n        align1 += seq1[j-1]\n        align2 += '-'\n        j -= 1\n    while i > 0:\n        align1 += '-'\n        align2 += seq2[i-1]\n        i -= 1\n    \n    align1 = align1[::-1]\n    align2 = align2[::-1]\n    \n    # Alignment score\n    align_score = 0\n    for i in range(len(align1)):\n        if align1[i] == '-' or align2[i] == '-':\n            align_score += gap_penalty\n        else:\n            align_score += match_score(align1[i], align2[i])\n    \n    return(align1, align2, align_score)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"1. **Run this on HEAGAWGHEE versus PAWHEAE and compare it to page 23 in lecture 5.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"# The two sequences on the lecture slides\nseq1 = \"HEAGAWGHEE\"\nseq2 = \"PAWHEAE\"\n\noutput1, output2, score1 = needleman_wunsch(seq1, seq2)\n\nprint(output1 + \"\\n\" + output2 + \"\\n\")\nprint(\"The alignment score is \" + str(score1) + \".\\n\")\nprint(\"That is a very low alignment score, which means the match is very poor.\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The result matches that on the lecture slides.\n\n<img src=\"https://i.imgur.com/3pRDOZs.png\" width=\"500\">"},{"metadata":{},"cell_type":"markdown","source":"2. **Match the protein sequence SALPQPTTPVSSFTSGSMLGRTDTALTNTYSAL with PSPTMEAVTSVEASTASHPHSTSSYFATTYYHLY.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"seq3 = \"SALPQPTTPVSSFTSGSMLGRTDTALTNTYSAL\"\nseq4 = \"PSPTMEAVTSVEASTASHPHSTSSYFATTYYHLY\"\n\noutput3, output4, score2 = needleman_wunsch(seq3, seq4)\n\nprint(output3 + \"\\n\" + output4 + \"\\n\")\nprint(\"The alignment score is \" + str(score2) + \".\\n\")\nprint(\"That is still a low alignment score, which means the match is poor.\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Task 2\n\n**Modify your program to implement the Smith-Waterman algorithm.**"},{"metadata":{},"cell_type":"markdown","source":"Compared to Needleman-Wunsch, the Smith-Waterman algorithm performs local alignment and the dynamic programming formula has one extra option, which is 0.\n\n<img src=\"https://i.imgur.com/iRfNRDA.png\" width=\"300\">\n\nIn addition to that, the backtracking starts from the cell that has the highest score.[](http://)"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Determine cell with highest score\ndef max_element(A):\n    r, (c, l) = max(map(lambda t: (t[0], max(enumerate(t[1]), key=lambda v: v[1])), enumerate(A)), key=lambda v: v[1][1])\n    return (r, c)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def smith_waterman(seq1, seq2):\n    \n    n = len(seq1)  \n    m = len(seq2)\n    \n    # Initiate empty scoring matrix\n    score = zeros(m+1, n+1)\n    \n    # Fill out first column\n    for i in range(0, m + 1):\n        score[i][0] = 0\n    \n    # Fill out first row\n    for j in range(0, n + 1):\n        score[0][j] = 0\n    \n    # Fill out all other values in the score matrix\n    for i in range(1, m + 1):\n        for j in range(1, n + 1):\n            \n            # Calculate the score by checking the top, left, and diagonal cells\n            match = score[i - 1][j - 1] + match_score(seq1[j-1], seq2[i-1])\n            delete = score[i - 1][j] + gap_penalty\n            insert = score[i][j - 1] + gap_penalty\n            \n            # Record the maximum score from the three possible scores calculated above\n            score[i][j] = max(match, delete, insert, 0)\n    \n    #print_matrix(score)\n\n    align1 = \"\"\n    align2 = \"\"\n    \n    # Start from the max cell in matrix\n    i, j = max_element(score)\n    \n    while score[i][j] > 0:\n        score_current = score[i][j]\n        score_diagonal = score[i-1][j-1]\n        score_up = score[i][j-1]\n        score_left = score[i-1][j]\n        \n        # Figure out which cell the score came from\n        if score_current == score_diagonal + match_score(seq1[j-1], seq2[i-1]):\n            align1 += seq1[j-1]\n            align2 += seq2[i-1]\n            i -= 1\n            j -= 1\n        elif score_current == score_up + gap_penalty:\n            align1 += seq1[j-1]\n            align2 += '-'\n            j -= 1\n        elif score_current == score_left + gap_penalty:\n            align1 += '-'\n            align2 += seq2[i-1]\n            i -= 1\n    \n    align1 = align1[::-1]\n    align2 = align2[::-1]\n    \n    # Alignment score\n    align_score = 0\n    for i in range(len(align1)):\n        if align1[i] == '-' or align2[i] == '-':\n            align_score += gap_penalty\n        else:\n            align_score += match_score(align1[i], align2[i])\n    \n    return(align1, align2, align_score)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"1. **Again, run this on HEAGAWGHEE versus PAWHEAE and compare it to page 5 in lecture 6.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"output5, output6, score3 = smith_waterman(seq1, seq2)\n\nprint(output5 + \"\\n\" + output6 + \"\\n\")\nprint(\"The alignment score is \" + str(score3) + \".\\n\")\nprint(\"That is a fairly high alignment score, which means the match is good.\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The result matches that on the lecture slides.\n\n<img src=\"https://i.imgur.com/hhZPL8d.png\" width=\"500\">"},{"metadata":{},"cell_type":"markdown","source":"2. **Find the best local match between MQNSHSGVNQLGGVFVNGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRY and TDDECHSGVNQLGGVFVGGRPLPDSTRQKIVELAHSGARPCDISRI.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"seq5 = \"MQNSHSGVNQLGGVFVNGRPLPDSTRQKIVELAHSGARPCDISRILQVSNGCVSKILGRY\"\nseq6 = \"TDDECHSGVNQLGGVFVGGRPLPDSTRQKIVELAHSGARPCDISRI\"\n\noutput7, output8, score4 = smith_waterman(seq5, seq6)\n\nprint(output7 + \"\\n\" + output8 + \"\\n\")\nprint(\"The alignment score is \" + str(score4) + \".\\n\")\nprint(\"That is a very high alignment score, which means the match is strong.\")","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Task 3\n**Test the BLAST algorithm.**\n1. **Download the Pax6 protein for the mouse by going to http://www.uniprot.org/uniprot/P63015 choose the \"Format\" tab and choose the FASTA (canonical) format**\n\n<img src=\"https://i.imgur.com/mDfQzU0.png\" width=\"700\">\n\n2. **Do the same for the eyeless protein for the fruit fly http://www.uniprot.org/uniprot/O96791**\n\n<img src=\"https://i.imgur.com/Mkc7XqV.png\" width=\"700\">\n\n3. **Perform a BLAST sequence comparison using the web service at https://blast.ncbi.nlm.nih.gov**\n\n<img src=\"https://i.imgur.com/750z3HR.png\" width=\"700\">\n\nThe mouse Pax6 gene and the fly eyeless gene are both associated with vision, which justifies why they are so remarkably similar in sequence, even though the hosts (animals) are very different. The original Pax gene probably evolved more than 500 million years ago. That animal's descendants evolved into all the diverse organisms alive today that carry eye-building genes. As these descendent lineages further evolved, the basic eye-building gene was modified in different ways in the different lineages, giving rise to the diversity of Pax6-like genes seen in modern animals. [1]\n\n[1] [https://evolution.berkeley.edu/evolibrary/article/1_0_0/eyes_10](https://evolution.berkeley.edu/evolibrary/article/1_0_0/eyes_10)"},{"metadata":{},"cell_type":"markdown","source":"# Task 4\n\n**Program the following HMM to generate CG rich regions.**\n\n![](https://i.imgur.com/ugIXoYP.png)"},{"metadata":{},"cell_type":"markdown","source":"Transition matrix:"},{"metadata":{"trusted":true},"cell_type":"code","source":"# The hidden states\nstates = [\"AT rich\", \"CG rich\"]\n\ntr_matrix = {\n    # The probabilities of switching states from AT rich\n    'AT rich': {'AT rich': 0.9998, 'CG rich': 0.0002},\n    # The probabilities of switching states from CG rich\n    'CG rich': {'AT rich': 0.0003, 'CG rich': 0.9997}\n}\n\n# Pretty print\nprint('        ', end=' ')\nfor state in states:\n    print(state, end='  ')\n    \nprint(\"\\nAT rich\", end='  ')\nfor state in tr_matrix['AT rich']:\n    print(tr_matrix['AT rich'][state], end='   ')\n    \nprint(\"\\nCG rich\", end='  ')\nfor state in tr_matrix['CG rich']:\n    print(tr_matrix['CG rich'][state], end='   ')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Emission matrix:"},{"metadata":{"trusted":true},"cell_type":"code","source":"observations = [\"A\", \"T\", \"C\", \"G\"]\n\nem_matrix = {\n    'AT rich': {'A': 0.2698, 'T': 0.3237, 'C': 0.2080, 'G': 0.1985},\n    'CG rich': {'A': 0.2459, 'T': 0.2079, 'C': 0.2478, 'G': 0.2984}\n}\n\n# Pretty print\nprint('        ', end=' ')\nfor obs in observations:\n    print(obs, end='        ')\n\nprint(\"\\nAT rich\", end='  ')\nfor obs in em_matrix['AT rich']:\n    print(em_matrix['AT rich'][obs], end='   ')\n\nprint(\"\\nCG rich\", end='  ')\nfor prob in em_matrix['CG rich']:\n    print(em_matrix['CG rich'][obs], end='   ')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The following code generates a sequence of length 10 using the given HMM."},{"metadata":{"trusted":true},"cell_type":"code","source":"# Assume the initial state is CG rich \ninitial_state = \"CG rich\"\nprobs = em_matrix['CG rich']\n\n# Determine the first observation in the sequence\ninitial_obs = np.random.choice(observations, size=None, replace=True, p=list(probs.values()))\nprint(\"Position 1: state = \" + initial_state + \"; observation = \" + initial_obs)\n\nsequence = list()\nseen = list()\n\n# Update the sequence and seen states\nsequence.append(initial_obs)\nseen.append(initial_state)\n\n# Determine the rest of the sequence\nfor i in range(9):\n    i = i + 2\n    \n    previous_state = seen[i-2]\n    \n    if previous_state == \"AT rich\":\n        current_state_p = tr_matrix['AT rich']\n        current_obs_p = em_matrix['AT rich']\n    else:\n        current_state_p = tr_matrix['CG rich']\n        current_obs_p = em_matrix['CG rich']\n        \n    current_state = np.random.choice(states, size=None, replace=True, p=list(current_state_p.values()))\n    current_obs = np.random.choice(observations, size=None, replace=True, p=list(current_obs_p.values()))\n    sequence.append(current_obs)\n    seen.append(current_state)\n    \n    print(\"Position \" + str(i) + \": state = \" + current_state + \n          \"; observation = \" + current_obs)\n    \nprint(\"\\nThe generated sequence is:\")\nprint(sequence)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Task 5\n\n**Write a viterbi algorithm for finding the most likely CG regions and find a way of drawing this.**"},{"metadata":{"trusted":true},"cell_type":"code","source":"# Creates a HMM configuration to run Viterbi on\nclass ViterbiHMM:\n    \n    def __init__(self, states, tr_matrix, em_matrix, st_prob, observations):\n\n        # Initialise HMM config\n        self.states = states\n        self.tr_matrix = tr_matrix\n        self.em_matrix = em_matrix\n        self.st_prob = st_prob\n        self.observations = observations\n        \n        # Forward pass dict\n        self.v = dict([[k,{}] for k in self.states])\n        self.sub_path = dict([[k, []] for k in self.states])\n        self.path = []\n    \n    def _v_i(self, t, csi=None, i=None):\n        \n        if t == 0:\n            if i == states[0]:\n                return 0\n            else:\n                # log(0) is undefined\n                return -np.inf\n            \n        e_i = self.em_matrix[i][csi]\n        \n        # We want to use log probabilities\n        v_js = [v[t-1] + np.log(self.tr_matrix[i][state]) for state, v in self.v.items()]\n        v_j = np.max(v_js)\n        v_ptr = np.argmax(v_js)\n        v_i = np.log(e_i) + v_j\n        \n        self.sub_path[i].append(v_ptr)\n        \n        return v_i\n    \n    # The forward pass algorithm\n    def forward(self):\n        \n        for state, v in self.v.items():\n            v[0] = self._v_i(0, csi='N/A', i=state)\n            \n        for t, k in enumerate(self.observations, 1):\n            for state, v in self.v.items():\n                v[t] = self._v_i(t, csi=k, i=state)\n    \n    # Backtracking\n    def backward(self):\n        \n        q_T = self.states[np.argmax([list(v.values())[-1] for v in self.v.values()])]\n        self.path.append(q_T)\n        \n        for i in range(len(self.sub_path[q_T])-1,0,-1):\n            q_T_prev = self.sub_path[q_T][i]\n            q_T_prev = self.states[q_T_prev]\n            self.path.append(q_T_prev)\n            q_T = q_T_prev\n            \n    # Run forward and backward\n    def runViterbi(self):\n        self.forward()\n        self.backward()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The Viterbi algorithm was run on the Dishonest Casino problem to test its functionality."},{"metadata":{"trusted":true},"cell_type":"code","source":"# Dishonest Casino config\nstates = ['Honest', 'Dishonest']\ntr_casino = {\n    'Honest': {'Honest': 0.9, 'Dishonest': 0.1},\n    'Dishonest': {'Honest': 0.1, 'Dishonest': 0.9}\n}\nem_casino = {\n    'Honest': {'1': 1/6, '2': 1/6, '3': 1/6, '4': 1/6, '5': 1/6, '6': 1/6},\n    'Dishonest': {'1': 1/10, '2': 1/10, '3': 1/10, '4': 1/10, '5': 1/10, '6': 1/2}\n}\nobservations = [obs for obs in \n                '54535254566666643656666356614166263656666211662113' + \n                '11155566351166565663466653642535666662541345464155']\n\n# Run Viterbi on Dishonest Casino\nv_casino = ViterbiHMM(states, tr_casino, em_casino, None, observations)\nv_casino.runViterbi()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Print the full path\ncasino_path = ['H' if thing == 'Honest' else 'D' for thing in v_casino.path]\n# print(casino_path)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"predicts = np.reshape(np.array([1 if thing == 'Honest' else 0 for thing in v_casino.path]), (-1, 10))\nfig, ax = plt.subplots(1,2,figsize=(20,10))\n\n# Path of predicted states\nfor i in range(predicts.shape[0]):\n    for j in range(predicts.shape[1]):\n        ax[0].text(i, j, casino_path[i+j*10], color='w')\n        \nax[0].imshow(predicts, cmap='jet')\n\n# Path of predicted observations\nfor i in range(predicts.shape[0]):\n    for j in range(predicts.shape[1]):\n        ax[1].text(i, j, v_casino.observations[i+j*10], color='w')\n        \nax[1].imshow(predicts, cmap='jet')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The figure on the left displays the path of state predictions made by the Viterbi algorithm (H/Red = Honest, D/Blue = Dishonest) and the one on the right displays the respective observations."},{"metadata":{},"cell_type":"markdown","source":"# Task 6\n**Run the Viterbi algorithm on the genome for the [escherichia phage lambda](https://www.ncbi.nlm.nih.gov/nuccore/215104?report=fasta).**"},{"metadata":{"trusted":true},"cell_type":"code","source":"# The config matches the initial one set up for the given HMM\nstates = [\"AT rich\", \"CG rich\"]\nobservations = []\n\n# Observations from the phase lambda genome file\nlines = open(\"../input/lambda/genome.txt\").read().splitlines()\nfor line in lines:\n   for char in line:\n      observations.append(char)\n\n# Run Viterbi on the genome\nv_genome = ViterbiHMM(states, tr_matrix, em_matrix, None, observations)\nv_genome.runViterbi()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Print the full path\n# print(['AT' if thing == 'AT rich' else 'CG' for thing in v_genome.path])","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Generating colour map for the genome's region predictions\npredicts = [1 if thing == 'AT rich' else 0 for thing in v_genome.path]\npredicts = np.array(predicts)\npredicts = np.concatenate((predicts, np.zeros(2)))\npredicts = np.reshape(predicts, (-1, 188))\n\nfig, ax = plt.subplots(figsize=(10,20))\nax.imshow(predicts, cmap='jet')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"The graph displays a visual representation of the full escherichia phage lambda genome and its distribution of AT and CG rich regions, where the respective regions have been coloured according to the following rule:\n\n* Red: Most likely AT rich region\n* Blue: Most likely CG rich region\n\nAs can be seen, CG rich region is maintained considerably more often than AT rich based on the predictions."},{"metadata":{},"cell_type":"markdown","source":"# Author\n\nSigne Rebassoo, sr2u17, ID: 29554896, University of Southampton"}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.6.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}